### ISSUE TITLE
[BUG]: `compare_models`unable to fit all available models after feature engineering on target and exog variables. 

### ISSUE THREAD

**Issue by ggjx22** [pycaret/pycaret#3788]
### pycaret version checks

- [X] I have checked that this issue has not already been reported [here](https://github.com/pycaret/pycaret/issues).

- [X] I have confirmed this bug exists on the [latest version](https://github.com/pycaret/pycaret/releases) of pycaret.

- [X] I have confirmed this bug exists on the master branch of pycaret (pip install -U git+https://github.com/pycaret/pycaret.git@master).

### Issue Description

Hi, first of all, thank you for providing us with such a wonderful library for automl tasks. I have been working specifically with `pycaret` Time Series module for a few weeks now. I am also trying to make use of `sktime` feature engineering transformers in hopes of better forecast results/accuracy.

I am experiencing some 'weird' behaviors during `compare_models` when I do feature engineering to produce more datetime and lag features. It could be the way I am setting the experiment during `setup` that is incorrect (the documentation is also outdated and not very comprehensive, some trial-and-error is required along the way). I should also mention that both `setup` and `compare_models` were able to run without errors. 

I realized that whenever I applied transformations to both target and exog variables, some models were not be fitted with error message `ERROR:Fit failed on ForecastingPipeline(steps=...`

#### ARIMA
```
ERROR:A constant trend was included in the model specification, but the `exog` data already contains a column of constants.
```

#### StatsForecastAutoARIMA
```
ERROR:xreg is rank deficient
```

#### LinearRegression, ElasticNet, Ridge, OrthogonalMatchingPursuit... (most of the regressor models)
```
ERROR:Input X contains NaN.
```

I am unsure why these errors are happening and the only way to avoid them is not to do any feature engineering at all and let `setup` handle feature engineering under the hood. But that is not really feasible as most of the time we want to create much more features.

#### output of `compare_models()`with feature engineering.
![Screen Shot 2023-10-21 at 00 04 44 AM](https://github.com/pycaret/pycaret/assets/134523967/b85db17e-1397-4ef3-948e-a318be6da52d)

#### output of `compare_models()`without any feature engineering.
![Screen Shot 2023-10-21 at 00 17 13 AM](https://github.com/pycaret/pycaret/assets/134523967/ea17e41a-3f1f-4b9a-b323-33c7db697e3d)

### Reproducible Example

```python
# load data
from pycaret.datasets import get_data
data = get_data("airline")

# customized a sliding window cross validation object
from sktime.transformations.series.summarize import SlidingWindowSplitter
import numpy as np 

splitter = SlidingWindowSplitter(
    fh=np.arange(1, 12+1),
    window_length=48,
    step_length=10
)

# create a custom function
from sklearn.metrics import mean_squared_log_error
from pycaret.time_series import add_metric

def rmsle(y_true, y_pred):
    return mean_squared_log_error(y_true
                        , y_pred
                        , squared=False)

# feature engineering on target variable
from sktime.transformations.series.summarize import WindowSummarizer

kwargs_target = {
    'lag_feature':{
        'lag':[1,2,3],
        'mean':[[0,1],[0,2],[0,3]],
        'std':[[0,1],[0,2],[0,3]]
    }
}

fe_target_rr = [WindowSummarizer(n_jobs=-1, truncate='bfill', **kwargs_target)]

# feature engineering on exog variables
from sktime.transformations.series.date import DateTimeFeatures
from sktime.transformations.series.fourier import FourierFeatures

fe_datetime = DateTimeFeatures(ts_freq='D', feature_scope='comprehensive', keep_original_columns=True)
fe_fourier = FourierFeatures(sp_list=[12,1], fourier_terms_list=[2,1], keep_original_columns=True)

fe_exog = [
    ('fe_datetime', fe_datetime),
    ('fe_fourier', fe_fourier)
]

# pycaret time series setup
from pycaret.time_series import TSForecastingExperiment

exp = TSForecastingExperiment()

expt = exp.setup(
  data=data,
  scale_exogenous='minmax',
  fe_target_rr=fe_target_rr,  # uncomment to disable feature engineering
  fe_exogenous=fe_exog,       # uncomment to disable feature engineering
  fold_strategy=splitter,
  seasonal_period=[12,1],
  sp_detection='auto',
  num_sps_to_use=-1,
  seasonality_type='auto',
  enforce_exogenous=False,
  n_jobs=-1,
  session_id=42,
  log_experiment=False,
)

expt.add_metric('rmsle', 'RMSLE', rmsle, greater_is_better=False)

# begin autoML and return the top 3 models with lowest RMSLE
best_models = expt.compare_models(
    sort='RMSLE',
    n_select=3,
    turbo=False,
    engine={'auto_arima':'statsforecast'}
)
```

### Expected Behavior

Most of the models from `models()`

### Actual Results

```python-traceback
# detailed error messages from logs.log

# arima
2023-10-20 20:49:05,296:ERROR:Fit failed on ForecastingPipeline(steps=[('transformer_exogenous',
                            TransformerPipeline(steps=[('fe_datetime',
                                                        DateTimeFeatures(feature_scope='comprehensive',
                                                                         keep_original_columns=True,
                                                                         ts_freq='D')),
                                                       ('fe_fourier',
                                                        FourierFeatures(fourier_terms_list=[2,
                                                                                            1],
                                                                        keep_original_columns=True,
                                                                        sp_list=[12,
                                                                                 1]))])),
                           ('forecaster',
                            TransformedTargetForecaster(steps=[('model',
                                                                ARIMA(seasonal_order=(0,
                                                                                      1,
                                                                                      0,
                                                                                      12)))]))])
2023-10-20 20:49:05,296:ERROR:A constant trend was included in the model specification, but the `exog` data already contains a column of constants.

# autoarima statsforecast
2023-10-20 20:49:06,524:ERROR:Fit failed on ForecastingPipeline(steps=[('transformer_exogenous',
                            TransformerPipeline(steps=[('fe_datetime',
                                                        DateTimeFeatures(feature_scope='comprehensive',
                                                                         keep_original_columns=True,
                                                                         ts_freq='D')),
                                                       ('fe_fourier',
                                                        FourierFeatures(fourier_terms_list=[2,
                                                                                            1],
                                                                        keep_original_columns=True,
                                                                        sp_list=[12,
                                                                                 1]))])),
                           ('forecaster',
                            TransformedTargetForecaster(steps=[('model',
                                                                StatsForecastAutoARIMA(sp=12))]))])
2023-10-20 20:49:06,524:ERROR:xreg is rank deficient

# regressor models. simillar messages for other regressor models which fit failed
2023-10-20 20:49:37,418:ERROR:Fit failed on ForecastingPipeline(steps=[('transformer_exogenous',
                            TransformerPipeline(steps=[('fe_datetime',
                                                        DateTimeFeatures(feature_scope='comprehensive',
                                                                         keep_original_columns=True,
                                                                         ts_freq='D')),
                                                       ('fe_fourier',
                                                        FourierFeatures(fourier_terms_list=[2,
                                                                                            1],
                                                                        keep_original_columns=True,
                                                                        sp_list=[12,
                                                                                 1]))])),
                           ('forecaster',
                            TransformedTargetForecaster(steps=[('model',
                                                                BaseCdsDtForecaster(fe_target_rr=[WindowSummarizer(lag_feature={'lag': [1,
                                                                                                                                        2,
                                                                                                                                        3],
                                                                                                                                'mean': [[0,
                                                                                                                                          1],
                                                                                                                                         [0,
                                                                                                                                          2],
                                                                                                                                         [0,
                                                                                                                                          3]],
                                                                                                                                'std': [[0,
                                                                                                                                         1],
                                                                                                                                        [0,
                                                                                                                                         2],
                                                                                                                                        [0,
                                                                                                                                         3]]},
                                                                                                                   truncate='bfill')],
                                                                                    regressor=LinearRegression(n_jobs=-1),
                                                                                    sp=12,
                                                                                    window_length=12))]))])
2023-10-20 20:49:37,419:ERROR:Input X contains NaN.
LinearRegression does not accept missing values encoded as NaN natively. For supervised learning, you might want to consider sklearn.ensemble.HistGradientBoostingClassifier and Regressor which accept missing values encoded as NaNs natively. Alternatively, it is possible to preprocess the data, for instance by using an imputer transformer in a pipeline or drop samples with missing values. See https://scikit-learn.org/stable/modules/impute.html You can find a list of all estimators that handle NaN values at the following page: https://scikit-learn.org/stable/modules/impute.html#estimators-that-handle-nan-values
```

### Installed Versions

```python
import pycaret
pycaret.__version__
```
`3.1.0`

**Comment by ggjx22** [comment]
hello @ngupta23, wanted to check in if you are looking into this issue? Thanks! :)

**Comment by ngupta23** [comment]
Can not help without the data (needed to debug the specific issue). But you can try adding an imputer at the end of your Feature Engineering to see if it helps fill in the missing values that it is complaining about.

**Comment by moezali1** [comment]
Closing as part of the 4.0 backlog cleanup — see #4177.

PyCaret 4.0 is a ground-up rewrite (OOP-only API, sklearn-native engine, Python 3.11+, sklearn ≥ 1.7) and almost all pre-4.0 issues either reference removed APIs (`setup()`, `plot_model`, `evaluate_model`, `interpret_model`, `automl`, the legacy time-series module, the Streamlit dashboard, …) or no longer reproduce on the new code path.

**If this still affects you in 4.0, please re-file as a new issue** with:

- the experiment constructor call (`Experiment(...)`),
- the operation that fails,
- the full traceback,
- your Python / sklearn / pycaret versions.

Resources for 4.0:

- Docs: https://pycaret.org/docs
- Migrate from 3.x: https://pycaret.org/docs/guides/migrating-from-3-x
- API reference: https://pycaret.org/reference
- FAQ: https://pycaret.org/docs/resources/faqs

Thanks for filing — and apologies for the long latency. The repo had drifted; we're starting clean.
